distributed/protocol/compression.py

```diff
@@ -214,3 +214,35 @@ def decompress(header, frames):
         compressions[c]["decompress"](frame)
         for c, frame in zip(header["compression"], frames)
     ]
+
+
+import collections.abc
+
+
+class TypeCompressor(collections.abc.MutableMapping):
+    def __init__(self, compressions):
+        self.storage = {}
+        self.compressions = compressions
+
+    def __setitem__(self, key, value):
+        compression_type = dask.config.get("distributed.comm.compression")
+        compression_type = "zlib"
+        if compression_type == "auto":
+            compression_type = default_compression
+        self.storage[key] = self.compressions[compression_type]["compress"](value)
```

Comment:
I suspect that this won't work for values that don't satisfy the buffer protocol.  

Compression functions generally consume bytes-like objects and return bytes-like objects.  Your test probably passes because Numpy arrays happen to satisfy this protocol, but other objects, like custom user classes, probably wouldn't.  

I recommend trying out your solution with a wider variety of workloads.  I suspect that it will have difficulty on many of them.

You will probably need to look into how Dask serializes objects to understand how to convert arbitrary Python objects into bytes-like objects.